### PR TITLE
Backend: Initiate repo cloning in preprocessor after execution request is received

### LINKED ISSUE #13: Test-preprocessor: Create a stub for tests preprocessing
* Backend makes POST request to preprocessor to clone repo (#62 )
* Preprocessor generates test IDs and POSTs them to backend
* Backend writes IDs into DB

### LINKED ISSUE #62: Backend: Initiate repo cloning in preprocessor after execution request is received

### PR DESCRIPTION
**Pull request by kentr0w**
### What's done:
* Added logic for backend to save project

* [x] closes #62

### CODE REVIEW

File: save-common/src/commonMain/kotlin/org/cqfn/save/entities/ProjectDto.kt [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+import org.cqfn.save.repository.GitRepository
+
+/**
+ * Data class of information about project
+ *
+ * @property project project
+ * @property gitRepository github repository
+ */
+@Serializable
+data class ProjectDto(
```
Comment (petertrr) [inline save-common/src/commonMain/kotlin/org/cqfn/save/entities/ProjectDto.kt:13]: Let's call it `ExecutionRequest` or something similar. In future it should support execution on standard suites as well as git repos.

File: save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/CloneRepositoryController.kt [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    private val webClient: WebClient,
+    ) {
+    private val log = LoggerFactory.getLogger(CloneRepositoryController::class.java)
+
+    /**
+     * Endpoint to save project
+     *
+     * @param projectDto information about project
+     */
+    @PostMapping(value = ["/cloneRepository"])
```
Comment (petertrr) [inline save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/CloneRepositoryController.kt:35]: I think this endpoint name is quite confusing; this is the main entry point from frontend, so it should be called somehing like `/submitExecutionRequest`

File: save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/CloneRepositoryController.kt [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    fun saveRepository(@RequestBody projectDto: ProjectDto): Mono<String> {
+        val project = projectDto.project
+        try {
+            projectService.saveProject(project)
```
Comment (petertrr) [inline save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/CloneRepositoryController.kt:39]: How will we handle case when project already exists in DB?

File: save-backend/src/test/kotlin/org/cqfn/save/backend/CloneRepoTest.kt [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        val project = Project("noname","1","1","1","1")
+        val gitRepo = GitRepository("1")
+        val projectDto = ProjectDto(project, gitRepo)
+        webClient.post()
+            .uri("/cloneRepository")
+            .contentType(MediaType.APPLICATION_JSON)
+            .body(BodyInserters.fromValue(projectDto))
+            .exchange()
+            .expectStatus()
+            .isEqualTo(HttpStatus.INTERNAL_SERVER_ERROR)  // because this post call preprocessor
```
Comment (petertrr) [inline save-backend/src/test/kotlin/org/cqfn/save/backend/CloneRepoTest.kt:39]: Can we have a mock of preprocessor here, so that we can safely test that all required requests are sent?

**Review by petertrr — CHANGES_REQUESTED** [review]

File: save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/CloneRepositoryController.kt (hunk 3 above)
Comment (kentr0w) [reply save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/CloneRepositoryController.kt:40]: In this case, project will be save.

File: save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/CloneRepositoryController.kt [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+     */
+    @PostMapping(value = ["/submitExecutionRequest"])
+    fun saveRepository(@RequestBody executionRequest: ExecutionRequest): Mono<String> {
+        val project = executionRequest.project
+        try {
+            projectService.saveProject(project)
+            log.info("Project ${project.id} saved")
+        } catch (exception: DataAccessException) {
+            log.error("Save error with ${project.id} project")
+            ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("Error to clone repo")
```
Comment (petertrr) [inline save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/CloneRepositoryController.kt:44]: Looks like this response isn't returned anywhere

File: save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/CloneRepositoryController.kt (hunk 3 above)
Comment (petertrr) [reply save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/CloneRepositoryController.kt:40]: As a new entry? That's not correct.

File: save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/CloneRepositoryController.kt (hunk 5 above)
Comment (kentr0w) [reply save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/CloneRepositoryController.kt:44]: Yes, fixed

File: save-backend/src/test/kotlin/org/cqfn/save/backend/CloneRepoTest.kt (hunk 4 above)
Comment (petertrr) [reply save-backend/src/test/kotlin/org/cqfn/save/backend/CloneRepoTest.kt:39]: Decided not to test different responses for now

File: save-backend/src/main/kotlin/org/cqfn/save/backend/configs/WebConfiguration.kt [hunk 6]
```diff
@@ -31,4 +32,14 @@ class WebConfiguration {
...
     }
+
+    /**
+     * @param builder
+     * @return web client
+     */
+    @Bean
+    fun webClient(builder: WebClient.Builder): WebClient =
+            builder
+                .baseUrl("http://preprocessor:5200")
```
Comment (petertrr) [inline save-backend/src/main/kotlin/org/cqfn/save/backend/configs/WebConfiguration.kt:43]: This should be set in `application.properties`

File: save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/CloneRepositoryController.kt [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+     */
+    @PostMapping(value = ["/submitExecutionRequest"])
+    fun saveRepository(@RequestBody executionRequest: ExecutionRequest): Mono<String> {
+        val project = executionRequest.project
+        try {
+            projectService.saveProject(project)
+            log.info("Project ${project.id} saved")
+        } catch (exception: DataAccessException) {
+            log.error("Save error with ${project.id} project")
+            return Mono.just("Error to clone repo")
```
Comment (petertrr) [inline save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/CloneRepositoryController.kt:42]: Which status code will be returned?

File: save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/CloneRepositoryController.kt [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    private val log = LoggerFactory.getLogger(CloneRepositoryController::class.java)
+
+    /**
+     * Endpoint to save project
+     *
+     * @param executionRequest information about project
+     * @return mono string
+     */
+    @PostMapping(value = ["/submitExecutionRequest"])
+    fun saveRepository(@RequestBody executionRequest: ExecutionRequest): Mono<String> {
```
Comment (petertrr) [inline save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/CloneRepositoryController.kt:35]: Pls rename the function as well

File: save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/CloneRepositoryController.kt [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+        val project = executionRequest.project
+        try {
+            projectService.saveProject(project)
+            log.info("Project ${project.id} saved")
+        } catch (exception: DataAccessException) {
+            log.error("Save error with ${project.id} project")
```
Comment (petertrr) [inline save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/CloneRepositoryController.kt:41]:
```suggestion
            log.error("Save error with ${project.id} project", exception)
```

File: save-backend/src/main/kotlin/org/cqfn/save/backend/repository/ProjectRepository.kt [hunk 10]
```diff
@@ -2,10 +2,8 @@ package org.cqfn.save.backend.repository
 
 import org.cqfn.save.entities.Project
 import org.springframework.data.jpa.repository.JpaRepository
-import org.springframework.stereotype.Repository
 
 /**
  * The repository of project entities
  */
-@Repository
```
Comment (petertrr) [inline save-backend/src/main/kotlin/org/cqfn/save/backend/repository/ProjectRepository.kt:10]: Why did you remove it?

File: save-common/src/commonMain/kotlin/org/cqfn/save/entities/Project.kt [hunk 11]
```diff
@@ -1,19 +1,54 @@
...
  * @property name
  * @property type type of the project, e.g. github or manually registered
  * @property url
  * @property description description of the project, may be absent
- * @property id
  */
 @Entity
-class Project(
+@Serializable
+data class Project(
```
Comment (petertrr) [inline save-common/src/commonMain/kotlin/org/cqfn/save/entities/Project.kt:14]: `@Entity` shouldn't be a data class

**Review by petertrr — CHANGES_REQUESTED** [review]

File: save-backend/src/main/kotlin/org/cqfn/save/backend/configs/WebConfiguration.kt [hunk 12]
```diff
@@ -31,4 +34,14 @@ class WebConfiguration {
             ok().contentType(TEXT_HTML).bodyValue(html)
         }
     }
+
+    /**
+     * @param builder
+     * @return web client
+     */
+    @Bean
+    fun webClient(builder: WebClient.Builder): WebClient =
```
Comment (petertrr) [inline save-backend/src/main/kotlin/org/cqfn/save/backend/configs/WebConfiguration.kt:45]: We should call it `preprocessorWebClient` because backend will query orchestrator too in the future

File: save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/CloneRepositoryController.kt [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+    fun submitExecutionRequest(@RequestBody executionRequest: ExecutionRequest): Mono<ResponseEntity<*>>? {
+        val project = executionRequest.project
+        try {
+            projectService.saveProject(project)
+            log.info("Project ${project.id} saved")
+        } catch (exception: DataAccessException) {
+            log.error("Save error with ${project.id} project", exception)
+            return Mono.just(ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("Error to clone repo"))
+        }
+        log.info("Starting to clone ${project.id} project")
```
Comment (petertrr) [inline save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/CloneRepositoryController.kt:46]:
```suggestion
        log.info("Sending request to preprocessor to start cloning ${project.id} project")
```

File: save-backend/src/main/kotlin/org/cqfn/save/backend/service/ProjectService.kt [hunk 14]
```diff
@@ -9,4 +10,14 @@ import org.springframework.stereotype.Service
...
  */
 @Service
-class ProjectService(private val projectRepository: ProjectRepository)
+class ProjectService(private val projectRepository: ProjectRepository) {
+    /**
+     * @param project
+     */
+    fun saveProject(project: Project) {
+        projectRepository
+            .findAll()
```
Comment (petertrr) [inline save-backend/src/main/kotlin/org/cqfn/save/backend/service/ProjectService.kt:19]: You can use [query by example](https://www.baeldung.com/spring-data-query-by-example) to retrieve only data of interest, and without filtering large amount of data in the service. You can also ignore id when defining the matcher, then custom `equals` will not be needed.

**Review by petertrr — CHANGES_REQUESTED** [review]

File: save-common/src/commonMain/kotlin/org/cqfn/save/entities/Project.kt (hunk 11 above)
Comment (petertrr) [reply save-common/src/commonMain/kotlin/org/cqfn/save/entities/Project.kt:14]: ? https://github.com/cqfn/save-cloud/pull/50/files/9f04ecdab44b901f02cbbe909f034cf101a7b99b..d6f889d2825b42a899deb19d61bf6a84ce146c39#r597583407

File: save-backend/src/test/kotlin/org/cqfn/save/backend/CloneRepoTest.kt [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+            webClient.post()
+                .uri("/submitExecutionRequest")
+                .contentType(MediaType.APPLICATION_JSON)
+                .body(BodyInserters.fromValue(it))
+                .exchange()
+                .expectStatus()
+                .isEqualTo(HttpStatus.INTERNAL_SERVER_ERROR)
+        }
+        val projects = projectRepository.findAll()
+        Assertions.assertTrue(projects.size == 2)
```
Comment (petertrr) [inline save-backend/src/test/kotlin/org/cqfn/save/backend/CloneRepoTest.kt:62]: Why is it 2 here, if you are passing the same project?

Comment (kentr0w) [reply save-backend/src/test/kotlin/org/cqfn/save/backend/CloneRepoTest.kt:62]: Because there is already one project

**Review by petertrr — APPROVED** [review]
lgtm
